### ISSUE TITLE
Ask how to incorporate it into an EDM model.

### ISSUE THREAD

**Issue by heeeeyuhang** [scxue/DM-NonUniform#2]
![question](https://github.com/user-attachments/assets/d466487b-4aae-4187-8ef0-c3ed9763858f)

**Comment by scxue** [comment]
Hi, to plug in code into an EDM model, you may consider substituting the NoiseSchedulerVP to the following NoiseSchedulerEDM :) and init an step_optim class with the EDM noise scheduler StepOptim(your_noise_schedule_edm_here) and modify the self.T to 80 in the following code.
```
class StepOptim(object):
    def __init__(self, ns):
        super().__init__()
        self.ns = ns
        self.T = 80.0 # t_T of diffusion sampling, for VP models, T=1.0; for EDM models, T=80.0
```

```
class NoiseScheduleEDM:
    def __init__(
            self,
            dtype=torch.float32,
        ):
        """Create a wrapper class for the forward SDE (EDM type).
        """
        self.T = 80
        self.t0 = 2e-3
        self.schedule = 'continuous'
        self.total_N = 1000

    def marginal_log_mean_coeff(self, t):
        """
        Compute log(alpha_t) of a given continuous-time label t in [0, T].
        """
        return torch.zeros_like(t)

    def marginal_alpha(self, t):
        """
        Compute alpha_t of a given continuous-time label t in [0, T].
        """
        return torch.ones_like(t)

    def marginal_std(self, t):
        """
        Compute sigma_t of a given continuous-time label t in [0, T].
        """
        return t

    def marginal_lambda(self, t):
        """
        Compute lambda_t = log(alpha_t) - log(sigma_t) of a given continuous-time label t in [0, T].
        """
        log_mean_coeff = self.marginal_log_mean_coeff(t)
        log_std = torch.log(self.marginal_std(t))
        return log_mean_coeff - log_std

    def inverse_lambda(self, lamb):
        """
        Compute the continuous-time label t in [0, T] of a given half-logSNR lambda_t.
        """
        return torch.exp(-lamb)
    
    def edm_sigma(self, t):
        return self.marginal_std(t)/self.marginal_alpha(t)
    
    def edm_inverse_sigma(self, edmsigma):
        sigma = edmsigma
        lambda_t = torch.log(1 / sigma)
        t = self.inverse_lambda(lambda_t)
        return t
```
